src/diagnostics.jl

```diff
@@ -55,25 +52,174 @@ function velocity_div!(grid::RegularCartesianGrid, u, v, w, div)
+    freq, int = validate_interval(frequency, interval)
+    profile = zeros(model.arch, model.grid, 1, 1, model.grid.Nz)
+    VerticalProfile(profile, field, freq, int, 0.0)
+end
+
+function run_diagnostic(model::Model, P::VerticalProfile{<:Array})
+    P.profile .= mean(data(P.field), dims=[1, 2])
+end
+
+@hascuda function run_diagnostic(model::Model, P::VerticalProfile{<:CuArray})
+    Nx, Ny, Nz = P.field.grid.Nx, P.field.grid.Ny, P.field.grid.Nz
+    sz = 2Nx * sizeof(eltype(P.profile))
+    @cuda threads=Nx blocks=(Ny, Nz) shmem=sz gpu_accumulate_xy!(P.profile, model.poisson_solver.storage, P.field.data, nothing, +)
+    P.profile /= (Nx*Ny)  # Normalize to get the mean from the sum.
+end
+
+####
+#### Horizontally averaged vertical profiles of products (e.g. covariances)
+####
+
+mutable struct ProductProfile{P, F1, F2, I, T} <: Diagnostic
+      profile :: P
+       field1 :: F1
+       field2 :: F2
+    frequency :: I
+     interval :: T
+     previous :: Float64
+end
+
+function ProductProfile(model, field1, field2; frequency=nothing, interval=nothing)
+    freq, int = validate_interval(frequency, interval)
+    profile = zeros(model.arch, model.grid, 1, 1, model.grid.Nz)
+    ProductProfile(profile, field1, field2, freq, int, 0.0)
+end
+
+function run_diagnostic(model::Model, P::ProductProfile{<:Array})
+    # This will allocate lots of memory but we mostly care about GPU performance for now...
+    P.profile .= mean(data(P.field1) .* data(P.field2), dims=[1, 2])
+end
+
+@hascuda function run_diagnostic(model::Model, P::ProductProfile{<:CuArray})
+    Nx, Ny, Nz = P.field1.grid.Nx, P.field1.grid.Ny, P.field1.grid.Nz
+    sz = 2Nx * sizeof(eltype(P.profile))
+    @cuda threads=Nx blocks=(Ny, Nz) shmem=sz gpu_accumulate_xy!(P.profile, model.poisson_solver.storage, P.field1.data, P.field2.data, +)
+    P.profile ./= (Nx*Ny)  # Normalize to get the mean from the sum.
+    return
+end
+
+####
+#### Velocity covariances
+####
+
+mutable struct VelocityCovarianceProfiles{UU, UV, UW, VV, VW, WW, I, T} <: Diagnostic
+    uu :: UU
+    uv :: UV
+    uw :: UW
+    vv :: VV
+    vw :: VW
+    ww :: WW
+    frequency :: I
+    interval  :: T
+    previous  :: Float64
+end
+
+function VelocityCovarianceProfiles(model; frequency=nothing, interval=nothing)
+    freq, int = validate_interval(frequency, interval)
+    
+    u, v, w = model.velocities.u, model.velocities.v, model.velocities.w
+    uu = ProductProfile(model, u, u; frequency=freq, interval=int)
+    uv = ProductProfile(model, u, v; frequency=freq, interval=int)
+    uw = ProductProfile(model, u, w; frequency=freq, interval=int)
+    vv = ProductProfile(model, v, v; frequency=freq, interval=int)
+    vw = ProductProfile(model, v, w; frequency=freq, interval=int)
+    ww = ProductProfile(model, w, w; frequency=freq, interval=int)
+
+    VelocityCovarianceProfiles(uu, uv, uw, vv, vw, ww, freq, int, 0.0)
+end
+
+function run_diagnostic(model::Model, vcp::VelocityCovarianceProfiles)

```

Comment:
See previous comment. The diagnostic functionality already loops over fields, so presumably each covariance could simply be its own diagnostic. I'm worried we are dedicating a lot of code to this functionality.